### PR TITLE
GG-34225 Fixed B+Tree corruption during concurrent replace and remove.

### PR DESCRIPTION
**Pull request by ibessonov**
https://ggsystems.atlassian.net/browse/GG-34225

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+    private static final short PAIR_INNER_IO = 30000;
+
+    /** */
+    private static final short PAIR_LEAF_IO = 30001;
+
+    /** */
+    protected static final int PAGE_SIZE = 512;
+
+    /** */
+    protected static final long MB = 1024 * 1024;
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:64]: You can use org.apache.ignite.internal.util.IgniteUtils#MB

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+        pageMem.start();
+
+        return pageMem;
+    }
+
+    /**
+     * @return Allocated meta page ID.
+     * @throws IgniteCheckedException If failed.
+     */
+    private FullPageId allocateMetaPage() throws IgniteCheckedException {
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:117]: Maybe reanme to **allocateIdxMetaPage**

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+        return new FullPageId(pageMem.allocatePage(CACHE_ID, PageIdAllocator.INDEX_PARTITION, PageIdAllocator.FLAG_IDX), CACHE_ID);
+    }
+
+    /**
+     * Short for {@code T2<Integer, Integer>}.
+     */
+    protected static class Pair extends T2<Integer, Integer> {
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:124]: U can use **org.apache.ignite.internal.util.lang.IgnitePair**

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+                pageMem,
+                null,
+                new AtomicLong(),
+                metaPageId,
+                reuseList,
+                new IOVersions<>(new TestPairInnerIO()),
+                new IOVersions<>(new TestPairLeafIO()),
+                PageIdAllocator.FLAG_IDX,
+                new FailureProcessor(new GridTestKernalContext(log)) {
+                    @Override public boolean process(FailureContext failureCtx) {
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:176]: /** {@inheritDoc} */

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 1 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:64]: True

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:117]: What does it have to do with index?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 3 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:124]: I'd still have to write <Integer> everywhere and I don't want it.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 4 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:176]: Do we write it in inner classes? It's not like I care, just wandering)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+                    tree.removex(new Pair(5, -1));
+                }
+                catch (IgniteCheckedException | BrokenBarrierException | InterruptedException e) {
+                    throw new IgniteException(e);
+                }
+            });
+
+            // Wait for both operations.
+            try {
+                putFut.get();
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:405]: Please add timeout

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+            }
+            finally {
+                remFut.get();
```
Comment (tkalkirill) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:408]: Please add timeout

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 5 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:405]: ok

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 6 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:408]: Done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 2 above)
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:117]: I see that the index partition appears here, but it doesn't seem to matter.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 3 above)
Comment (tkalkirill) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:124]: Ок

File: modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/database/BPlusTreeReplaceRemoveRaceTest.java:117]: Yes, any partition could be used there

**Review by tkalkirill — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 7]
```diff
@@ -617,6 +604,28 @@ else if (needBackIfRouting) {
         }
     }
 
+    /**
+     * Page handler that adds the page to the tail of {@link Update} object.
+     * Results in {@link Result#FOUND} if added to tail successfully.
+     * Results in {@link Result#RETRY} if triangle invariant is violated.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:610]: Is this "triangle invariant" some common knowledge? I could find any explanation in this class

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 8]
```diff
@@ -2447,16 +2457,20 @@ private T doPut(T row, boolean needOld) throws IgniteCheckedException {
                         continue;
 
                     case FOUND:
-                        // We may need to insert split key into upper level here.
+                        // We may need to perform an inner replace in the upper level.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2460]:
```suggestion
                        // We may need to perform an inner replace on the upper level.
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 9]
```diff
@@ -2447,16 +2457,20 @@ private T doPut(T row, boolean needOld) throws IgniteCheckedException {
...
                         if (!p.isFinished()) {
-                            // It must be impossible to have an insert higher than the current root,
-                            // because we are making decision about creating new root while keeping
-                            // write lock on current root, so it can't concurrently change.
-                            assert p.btmLvl <= getRootLevel();
+                            res = p.finishTail();
 
-                            checkInterrupted();
+                            // If not found, then the tree grew beyond our call stack -> retry from the actual root.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2464]: what do you mean by `grew beyond our call stack`?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 10]
```diff
@@ -3617,55 +3611,108 @@ public T find() throws IgniteCheckedException {
...
+            Result res = finishTail();
 
-            // If we can get full row from the inner page, we have to replace it with the new one. On the way down
-            // we can not miss inner key even in presence of concurrent operations because of `triangle` invariant +
-            // concurrent inner replace handling by retrying from root.
-            if (canGetRowFromInner && needReplaceInner == FALSE)
-                needReplaceInner = TRUE;
+            // Add this page to the tail if inner replace has not happened.
+            if (res == NOT_FOUND) {
+                // Set forward id to check triangle invariand under the write-lock.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:3639]:
```suggestion
                // Set forward id to check the triangle invariant under the write-lock.
```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 11]
```diff
@@ -3617,55 +3611,108 @@ public T find() throws IgniteCheckedException {
...
          * @param tailPage Tail page pointer.
          * @param tailPageAddr Tail page address
+         * @param io Tail page IO.
+         * @param lvl Tail page level.
          */
-        private void tail(long tailId, long tailPage, long tailPageAddr) {
-            assert (tailId == NULL_PAGE_ID) == (tailPage == NULL_PAGE);
-            assert (tailPage == NULL_PAGE) == (tailPageAddr == NULL_PAGE_ADDRESS);
+        private void setTailForSplit(long tailId, long tailPage, long tailPageAddr, BPlusIO<L> io, int lvl) {
+            assert tailId != 0L && tailPage != 0L && tailPageAddr != 0L;
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:3710]: why did you remove the constants?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 7 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:610]: There's no explicit explanation, but it can be derived from most places when it is checked. If there's a good place for it - I'll add a comment

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 9 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2464]: This is a copy from already existing comment. Nonetheless, it means that tree height has increased while we were attempting this put operation.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 11 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:3710]: 1. They were used here and in 2 (I think) more places where I replaced them with "foo = 0L".
2. Every other place in the code uses 0L explicitly. So, we should either used such constants everywhere or not using them at all

Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:3710]: Is it possible to use them everywhere?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 9 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2464]: Can you update the comment, please?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 7 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:610]: I think the class' javadoc might be a suitable place

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 11 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:3710]: That's too many small changes for a bug fix

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 9 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java (hunk 11 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:3710]: okay =(

**Review by sashapolo — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java [hunk 12]
```diff
@@ -2027,25 +2035,27 @@ private Result invokeDown(final Invoke x, final long pageId, final long backId,
...
 
                         return res;
 
                     case NOT_FOUND:
                         if (lvl == 0)
                             x.invokeClosure();
 
+                        // Level must be equal to bottom level. This is the place when we would insert values into
+                        // parent nodes during splits.
+                        assert lvl == (x.isPut() ? ((Put)x.op).btmLvl : 0) : "NOT_FOUND on the wrong level: lvl=" + lvl;
```
Comment (sergey-chugunov-1985) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2050]: How about adding more information to the assert? Would `btmLvl` be useful in case this invariant is broken?

Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java:2050]: This was more about documenting the code, but sure, I'll add bottom level to the output.
